### DISCUSSION TITLE
subgroupAll failing

### DISCUSSION THREAD

**Discussion by fhoenig** [KhronosGroup/MoltenVK#1826] [General]
subgroupAll gives me the following error when running on iOS:

[mvk-error] VK_ERROR_INITIALIZATION_FAILED: Compute pipeline compile failed (Error code 3):
Encountered unlowered function call to air.quad_all.

Shouldn't it translate to simd_all ?

**Comment by cdavis5e** [comment]
It depends on your hardware. Older SoCs only support quad operations. Newer ones support `simd_*`, and we should be using them on those; if they're not, you should probably convert this to a bug.

**Reply by fhoenig** [reply]
Thank you for responding Chip.

I am actually a bit confused about the terminology and distinction in the MSL spec. My mental model of the GPU and intention w the instruction is to have the predicate be true across a simd set of threads. quad sounds a bit like fragment processing specialization. Gonna look at this more deeply tomorrow but do you know the difference?

**Reply by cdavis5e** [reply]
An SIMD-group is simply a group of threads that is run in lockstep. SIMD-group size is variable; it can be any size supported by the hardware. Most hardware only supports 32 threads; earlier AMD GPUs only support 64 threads, and later ones also support 32 threads. I'm not entirely certain, but I believe that Apple GPUs from A13 on support any power-of-two SIMD-group size between 4 and 32--the docs indicate that the SIMD-group size can depend on GPU register pressure.

A quad-group, on the other hand, is like an SIMD-group in that it is a group of threads run in lockstep, but it is *always* of size 4. There may thus be multiple such quad-groups inside an SIMD-group. The Vulkan WG is supposed to be working on `Quad`-scoped subgroup operations that work on these quad-groups; they would correspond to the `quad_*` operations in Metal. (Some hardware even supports fragment-style differentiation in compute shaders (i.e. `dfdx()`, `dfdy()`, and `fwidth()`), which doesn't even make sense unless they run in some sort of quad-group. Note that AFAIK, Metal does not yet expose this; but see, e.g. the Vulkan extension `VK_NV_compute_shader_derivatives`.)

It used to be that the `simd_*` operations were only available on macOS, *in addition* to the `quad_*` operations; on iOS, only the `quad_*` operations were available. I assume this is because older Apple SoCs (A12 and below) didn't support any SIMD-group larger than 4. Later, in iOS 13, and only on GPUs supporting at least the Apple 6 family (i.e. A13 and up), Apple provided the `simd_*` operations there as well.

**Reply by fhoenig** [reply]
So you are saying `subgroupAll` mapping to either simd_ or quad_ depending on hardware is expected behavior of the cross compilation chain?

**Reply by cdavis5e** [reply]
Yes.

**Reply by fhoenig** [reply]
That is odd.
Because if simd_ intrinsics can be at variable size from 4 to 32, that includes 4-wide. Switching the type of instrinsic for older hardware then is an unnecessary semantic violation of sorts. 

Btw, where does that happen? in spir-v to msl or in the msl compiler itself?

**Reply by cdavis5e** [reply]
> That is odd. Because if simd_ intrinsics can be at variable size from 4 to 32, that includes 4-wide. Switching the type of instrinsic for older hardware then is an unnecessary semantic violation of sorts.

We are forced to do this. The Metal compiler backend for A12 and under refuses to compile any `simd_*` functions; it fails with an ICE.

> Btw, where does that happen? in spir-v to msl or in the msl compiler itself?

In SPIRV-Cross, so in the translation from SPIR-V to MSL.

**Reply by fhoenig** [reply]
But can the A12 and under run "waves" larger than 4? I think it might be. If this is true, then this "fallback" is just wrong or confusing in my opinion.

Anyways, not meaning to complain here. MoltenVK is awesome work and works like a charm otherwise, even for our "esoteric" tech.
